File: content/problems/0684-redundant-connection/solutions/union-find/Solution.java
class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        int[] parent = new int[edges.length + 1]; for (int node = 1; node < parent.length; node++) parent[node] = node;
        for (int[] edge : edges) { int first = find(parent, edge[0]), second = find(parent, edge[1]); if (first == second) return edge; parent[first] = second; }
        return new int[0];
    }
    private int find(int[] parent, int node) { if (parent[node] != node) parent[node] = find(parent, parent[node]); return parent[node]; }
}

File: content/problems/0684-redundant-connection/solutions/union-find/solution.cpp
class Solution {
    int find(vector<int>& parent, int node) { return parent[node] == node ? node : parent[node] = find(parent, parent[node]); }
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        vector<int> parent(edges.size() + 1); iota(parent.begin(), parent.end(), 0);
        for (auto& edge : edges) { int first = find(parent, edge[0]), second = find(parent, edge[1]); if (first == second) return edge; parent[first] = second; }
        return {};
    }
};
